File: murmurServer/app/src/main/java/org/murmurServer/grammar/Protocol.java
package org.murmurServer.grammar;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Protocol {
    // Définitions standards :
    private final String RX_CHIFFRE = "[0-9]";
    private final String RX_LETTRE = "[a-zA-Z]";
    private final String RX_LETTRE_CHIFFRE = String.format("[%s%s]", RX_LETTRE, RX_CHIFFRE);
    private final String RX_CARACTERE_IMPRIMABLE = "[\\x20-\\xff]";
    private final String RX_CRLF = "\\x0d\\x0a"; // OU => \\r\\n
    private final String RX_SYMBOLE = "[\\x21-\\x2f\\x3a-\\x40\\x5b-\\x60]";
    private final String RX_ESP = "\\x20";
    private final String RX_DOMAINE = String.format("[%s%s]{5,200}", RX_LETTRE_CHIFFRE,"\\x2e");
    private final String RX_PORT = String.format("%s{1,5}", RX_CHIFFRE);
    private final String RX_ROUND = String.format("%s{2}", RX_CHIFFRE);
    private final String RX_BCRYPT_HASH = String.format("\\$2b\\$%s\\$[%s%s]{1,70}", RX_ROUND, RX_LETTRE_CHIFFRE, RX_SYMBOLE);
    private final String RX_SHA3_HEX = String.format("%s{30,200}", RX_LETTRE_CHIFFRE);
    private final String RX_SALT_SIZE = String.format("%s{2}", RX_CHIFFRE);
    private final String RX_RANDOM22 = String.format("[%s%s]{22}", RX_LETTRE_CHIFFRE, RX_SYMBOLE);
    private final String RX_SALT = RX_RANDOM22;
    private final String RX_MESSAGE = String.format("%s{1,200}", RX_CARACTERE_IMPRIMABLE);
    private final String RX_MESSAGE_INTERNE = String.format("%s{1,500}", RX_CARACTERE_IMPRIMABLE);
    private final String RX_NOM_UTILISATEUR = String.format("%s{5,20}", RX_LETTRE_CHIFFRE);
    private final String RX_TAG = String.format("#%s{5,20}", RX_LETTRE_CHIFFRE);
    private final String RX_NOM_DOMAINE = String.format("%s@%s", RX_NOM_UTILISATEUR, RX_DOMAINE);
    private final String RX_TAG_DOMAINE = String.format("%s@%s", RX_TAG, RX_DOMAINE);
    private final String RX_ID_DOMAINE = String.format("%s{1,5}@%s", RX_CHIFFRE, RX_DOMAINE);

    // Echanges entre le client et le Murmur Server :
    private final String RX_HELLO = String.format("(HELLO)%s(%s)%s(%s)%s", RX_ESP, RX_DOMAINE, RX_ESP, RX_RANDOM22, RX_CRLF);
    private final String RX_CONNECT = String.format("(CONNECT)%s(%s)%s", RX_ESP, RX_NOM_UTILISATEUR, RX_CRLF);
    private final String RX_PARAM = String.format("(PARAM)%s(%s)%s(%s)%s", RX_ESP, RX_ROUND, RX_ESP, RX_SALT, RX_CRLF);
    private final String RX_CONFIRM = String.format("(CONFIRM)%s(%s)%s", RX_ESP, RX_SHA3_HEX, RX_CRLF);
    private final String RX_REGISTER = String.format("(REGISTER)%s(%s)%s(%s)%s(%s)%s", RX_ESP, RX_NOM_UTILISATEUR, RX_ESP, RX_SALT_SIZE, RX_ESP, RX_BCRYPT_HASH, RX_CRLF);
    private final String RX_FOLLOW = String.format("(FOLLOW)%s(%s|%s)%s", RX_ESP, RX_NOM_DOMAINE, RX_TAG_DOMAINE, RX_CRLF);
    private final String RX_MSG = String.format("(MSG)%s(%s)%s", RX_ESP, RX_MESSAGE, RX_CRLF);
    private final String RX_MSGS = String.format("(MSGS)%s(%s)%s(%s)%s", RX_ESP, RX_NOM_DOMAINE, RX_ESP, RX_MESSAGE, RX_CRLF);
    private final String RX_OK = String.format("(\\+OK)%s(%s)%s", RX_ESP, RX_MESSAGE, RX_CRLF);
    private final String RX_ERROR = String.format("(\\-ERR)%s(%s)%s", RX_ESP, RX_MESSAGE, RX_CRLF);
    private final String RX_DISCONNECT = String.format("(DISCONNECT)%s", RX_CRLF);

    // Echanges multicast entre Murmur Server et Murmur Relay :
    private final String RX_ECHO = String.format("(ECHO)%s(%s)%s(%s)%s", RX_ESP, RX_PORT, RX_ESP, RX_DOMAINE, RX_CRLF);

    // Echanges unicast entre Murmur Server et Murmur Relay :
    private final String RX_SEND = String.format("(SEND)%s(%s)%s(%s)%s(%s|%s)%s(%s)%s", RX_ESP, RX_ID_DOMAINE, RX_ESP, RX_NOM_DOMAINE, RX_ESP, RX_NOM_DOMAINE, RX_TAG_DOMAINE, RX_ESP, RX_MESSAGE_INTERNE, RX_CRLF);

    // Messages pour la construction :
    private final String HELLO = "HELLO <domaine> <random22>\r\n";
    private final String CONNECT = "CONNECT <nom-utilisateur>\r\n";
    private final String PARAM = "PARAM <round> <salt>\r\n";
    private final String CONFIRM = "CONFIRM <sha3>\r\n";
    private final String REGISTER = "REGISTER <nom-utilisateur> <salt-size> <bcrypt-hash>\r\n";
    private final String FOLLOW = "FOLLOW <nom-ou-tag-domaine>\r\n";
    private final String MSG = "MSG <message>\r\n";
    private final String MSGS = "MSGS <nom-domaine> <message>\r\n";
    private final String OK = "+OK <message>\r\n";
    private final String ERROR = "-ERR <message>\r\n";
    private final String DISCONNECT = "DISCONNECT\r\n";
    private final String ECHO = "ECHO <port> <domaine>\r\n";
    private final String SEND = "SEND <id-domaine> <nom-domaine> <nom-ou-tag-domaine> <message-interne>\r\n";

    public String buildHello(String domaine,String random22) {
        String helloMessage = HELLO.replaceAll("<domaine>",domaine).replaceAll("<random22>",random22);
        if(helloMessage.matches(RX_HELLO)) {
            return helloMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct arguments for the construction of the HELLO message");
        }
    }

    public String buildConnect(String username) {
        String connectMessage = CONNECT.replaceAll("<nom-utilisateur>",username);
        if(connectMessage.matches(RX_CONNECT)) {
            return connectMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct argument for the construction of the CONNECT message");
        }
    }

    public String buildParam(String round,String salt) {
        String paramMessage = PARAM.replaceAll("<round>",round).replaceAll("<salt>",salt);
        if(paramMessage.matches(RX_PARAM)) {
            return paramMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct arguments for the construction of the PARAM message");
        }
    }

    public String buildConfirm(String sha3) {
        String confirmMessage = CONFIRM.replaceAll("<sha3>",sha3);
        if(confirmMessage.matches(RX_CONFIRM)) {
            return confirmMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct argument for the construction of the CONFIRM message");
        }
    }

    public String buildRegister(String username,String saltSize,String bcryptHash) {
        String registerMessage = REGISTER.replaceAll("<nom-utilisateur>",username).replaceAll("<salt-size>",saltSize).replaceAll("<bcrypt-hash>",bcryptHash);
        if(registerMessage.matches(RX_REGISTER)) {
            return registerMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct arguments for the construction of the REGISTER message");
        }
    }

    public String buildFollow(String nameOrTagDomain) {
        String followMessage = FOLLOW.replaceAll("<nom-ou-tag-domaine>",nameOrTagDomain);
        if(followMessage.matches(RX_FOLLOW)) {
            return followMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct argument for the construction of the FOLLOW message");
        }
    }

    public String buildMsg(String message) {
        String msgMessage = MSG.replaceAll("<message>",message);
        if(msgMessage.matches(RX_MSG)) {
            return msgMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct argument for the construction of the MSG message");
        }
    }

    public String buildMsgs(String domainName,String message) {
        String msgsMessage = MSGS.replaceAll("<nom-domaine>",domainName).replaceAll("<message>",message);
        if(msgsMessage.matches(RX_MSGS)) {
            return msgsMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct arguments for the construction of the MSGS message");
        }
    }

    public String buildOk(String message) {
        String okMessage = OK.replaceAll("<message>",message);
        if(okMessage.matches(RX_OK)) {
            return okMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct argument for the construction of the OK message");
        }
    }

    public String buildError(String message) {
        String errorMessage = ERROR.replaceAll("<message>",message);
        if(errorMessage.matches(RX_ERROR)) {
            return errorMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct argument for the construction of the ERROR message");
        }
    }

    public String buildDisconnect() {
        return DISCONNECT;
    }

    public String buildEcho(String port,String domain) {
        String echoMessage = ECHO.replace("<port>",port).replaceAll("<domaine>",domain);
        if(echoMessage.matches(RX_ECHO)) {
            return echoMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct arguments for the construction of the ECHO message");
        }
    }

    public String buildSend(String idDomain,String domainName,String nameOrTagDomain,String internalMessage) {
        String sendMessage = SEND.replaceAll("<id-domaine>",idDomain).replaceAll("<nom-domaine>",domainName).replaceAll("<nom-ou-tag-domaine>",nameOrTagDomain).replaceAll("<message-interne>",internalMessage);
        if(sendMessage.matches(RX_SEND)) {
            return sendMessage;
        } else {
            throw new IllegalArgumentException("Please provide correct arguments for the construction of the SEND message");
        }
    }

    public String[] verifyMessage(String message) {
        if (message != null) {
            switch (message.split(" ")[0].trim()) {
                case "HELLO" -> {
                    return verifyHello(message);
                }
                case "CONNECT" -> {
                    return verifyConnect(message);
                }
                case "PARAM" -> {
                    return verifyParam(message);
                }
                case "CONFIRM" -> {
                    return verifyConfirm(message);
                }
                case "REGISTER" -> {
                    return verifyRegister(message);
                }
                case "FOLLOW" -> {
                    return verifyFollow(message);
                }
                case "MSG" -> {
                    return verifyMsg(message);
                }
                case "MSGS" -> {
                    return verifyMsgs(message);
                }
                case "+OK" -> {
                    return verifyOk(message);
                }
                case "-ERR" -> {
                    return verifyError(message);
                }
                case "DISCONNECT" -> {
                    return verifyDisconnect(message);
                }
                case "ECHO" -> {
                    return verifyEcho(message);
                }
                case "SEND" -> {
                    return verifySend(message);
                }
                default -> {
                    return new String[]{"-ERR", "The message does not exist in this protocol"};
                }
            }
        } else {
            return new String[]{"-ERR","The message must be NOT NULL"};
        }
    }

    private String[] verifyHello(String message) {
        Pattern p = Pattern.compile(RX_HELLO);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2),m.group(3)};
        }
        return new String[]{"-ERR","This HELLO message is not correctly formatted"};
    }

    private String[] verifyConnect(String message) {
        Pattern p = Pattern.compile(RX_CONNECT);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2)};
        }
        return new String[]{"-ERR","This CONNECT message is not correctly formatted"};
    }

    private String[] verifyParam(String message) {
        Pattern p = Pattern.compile(RX_PARAM);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2),m.group(3)};
        }
        return new String[]{"-ERR","This PARAM message is not correctly formatted"};
    }

    private String[] verifyConfirm(String message) {
        Pattern p = Pattern.compile(RX_CONFIRM);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2)};
        }
        return new String[]{"-ERR","This CONFIRM message is not correctly formatted"};
    }

    private String[] verifyRegister(String message) {
        Pattern p = Pattern.compile(RX_REGISTER);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2),m.group(3),m.group(4)};
        }
        return new String[]{"-ERR","This REGISTER message is not correctly formatted"};
    }

    private String[] verifyFollow(String message) {
        Pattern p = Pattern.compile(RX_FOLLOW);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2)};
        }
        return new String[]{"-ERR","This FOLLOW message is not correctly formatted"};
    }

    private String[] verifyMsg(String message) {
        Pattern p = Pattern.compile(RX_MSG);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2)};
        }
        return new String[]{"-ERR","This MSG message is not correctly formatted"};
    }

    private String[] verifyMsgs(String message) {
        Pattern p = Pattern.compile(RX_MSGS);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2),m.group(3)};
        }
        return new String[]{"-ERR","This MSGS message is not correctly formatted"};
    }

    private String[] verifyOk(String message) {
        Pattern p = Pattern.compile(RX_OK);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2)};
        }
        return new String[]{"-ERR","This OK message is not correctly formatted"};
    }

    private String[] verifyError(String message) {
        Pattern p = Pattern.compile(RX_ERROR);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2)};
        }
        return new String[]{"-ERR","This ERROR message is not correctly formatted"};
    }

    private String[] verifyDisconnect(String message) {
        Pattern p = Pattern.compile(RX_DISCONNECT);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1)};
        }
        return new String[]{"-ERR","This DISCONNECT message is not correctly formatted"};
    }

    private String[] verifyEcho(String message) {
        Pattern p = Pattern.compile(RX_ECHO);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2),m.group(3)};
        }
        return new String[]{"-ERR","This ECHO message is not correctly formatted"};
    }

    private String[] verifySend(String message) {
        Pattern p = Pattern.compile(RX_SEND);
        Matcher m = p.matcher(message);

        if(m.matches()) {
            return new String[]{m.group(1),m.group(2),m.group(3),m.group(4),m.group(5)};
        }
        return new String[]{"-ERR","This SEND message is not correctly formatted"};
    }
}

File: murmur_relay/src/grammar/protocol.rs
use std::string::String;
use regex::Regex;

#[derive(Clone)]
pub struct Protocol {
    // Définitions standards :
    rx_chiffre: String,
    rx_lettre: String,
    rx_lettre_chiffre: String,
    rx_caractere_imprimable: String,
    rx_crlf: String,
    rx_symbole: String,
    rx_esp: String,
    rx_domaine: String,
    rx_port: String,
    rx_round: String,
    rx_bcrypt_hash: String,
    rx_sha3_hex: String,
    rx_salt_size: String,
    rx_random22: String,
    rx_salt: String,
    rx_message: String,
    rx_message_interne: String,
    rx_nom_utilisateur: String,
    rx_tag: String,
    rx_nom_domaine: String,
    rx_tag_domaine: String,
    rx_id_domaine: String,

    // Echanges entre le client et le Murmur Server :
    rx_hello: String,
    rx_connect: String,
    rx_param: String,
    rx_confirm: String,
    rx_register: String,
    rx_follow: String,
    rx_msg: String,
    rx_msgs: String,
    rx_ok: String,
    rx_error: String,
    rx_disconnect: String,

    // Echanges multicast entre Murmur Server et Murmur Relay :
    rx_echo: String,

    // Echanges unicast entre Murmur Server et Murmur Relay :
    rx_send: String,

    hello: String,
    connect: String,
    param: String,
    confirm: String,
    register: String,
    follow: String,
    msg: String,
    msgs: String,
    ok: String,
    error: String,
    disconnect: String,
    echo: String,
    send: String,
}

impl Protocol {
    pub fn new() -> Self {
        let rx_chiffre = "[0-9]".to_string();
        let rx_lettre = "[a-zA-Z]".to_string();
        let rx_lettre_chiffre = format!("[{}{}]", rx_lettre, rx_chiffre).to_string();
        let rx_caractere_imprimable = "[\\x20-\\xff]".to_string();
        let rx_crlf = "\\x0d\\x0a".to_string();
        let rx_symbole = "[\\x21-\\x2f\\x3a-\\x40\\x5b-\\x60]".to_string();
        let rx_esp = "\\x20".to_string();
        let rx_domaine = format!("[{}{}]{{5,200}}", rx_lettre_chiffre, "\\x2e").to_string();
        let rx_port = format!("{}{{1,5}}", rx_chiffre).to_string();
        let rx_round = format!("{}{{2}}", rx_chiffre).to_string();
        let rx_bcrypt_hash = format!("\\$2b\\${}\\[{}{}\\]{{1,70}}", rx_round, rx_lettre_chiffre, rx_symbole).to_string();
        let rx_sha3_hex = format!("{}{{30,200}}", rx_lettre_chiffre).to_string();
        let rx_salt_size = format!("{}{{2}}", rx_chiffre).to_string();
        let rx_random22 = format!("[{}{}]{{22}}", rx_lettre_chiffre, rx_symbole).to_string();
        let rx_salt = rx_random22.clone();
        let rx_message = format!("{}{{1,200}}", rx_caractere_imprimable).to_string();
        let rx_message_interne = format!("{}{{1,500}}", rx_caractere_imprimable).to_string();
        let rx_nom_utilisateur = format!("{}{{5,20}}", rx_lettre_chiffre).to_string();
        let rx_tag = format!("#{}{{5,20}}", rx_lettre_chiffre).to_string();
        let rx_nom_domaine = format!("{}@{}", rx_nom_utilisateur, rx_domaine).to_string();
        let rx_tag_domaine = format!("{}@{}", rx_tag, rx_domaine).to_string();
        let rx_id_domaine = format!("{}{{1,5}}@{}", rx_chiffre, rx_domaine).to_string();

        let rx_hello = format!(r"(HELLO){}({}){}({})({}){{0,1}}", rx_esp, rx_domaine, rx_esp, rx_random22, rx_crlf).to_string();
        let rx_connect = format!("(CONNECT){}({}){}({})({}){{0,1}}", rx_esp, rx_nom_utilisateur, rx_esp, rx_sha3_hex, rx_crlf).to_string();
        let rx_param = format!(r"(PARAM){}({}){}({})({}){{0,1}}", rx_esp, rx_round, rx_esp, rx_salt, rx_crlf).to_string();
        let rx_confirm = format!(r"(CONFIRM){}({})({}){{0,1}}", rx_esp, rx_sha3_hex, rx_crlf).to_string();
        let rx_register = format!(r"(REGISTER){}({}){}({}){}({})({}){{0,1}}", rx_esp, rx_nom_utilisateur, rx_esp, rx_salt_size, rx_esp, rx_bcrypt_hash, rx_crlf).to_string();
        let rx_follow = format!(r"(FOLLOW){}({}|{})({}){{0,1}}", rx_esp, rx_nom_domaine, rx_tag_domaine, rx_crlf).to_string();
        let rx_msg = format!(r"(MSG){}({})({}){{0,1}}", rx_esp, rx_message, rx_crlf).to_string();
        let rx_msgs = format!(r"(MSGS){}({}){}({})({}){{0,1}}", rx_esp, rx_nom_domaine, rx_esp, rx_message, rx_crlf).to_string();
        let rx_ok = format!(r"(\\+OK){}({})({}){{0,1}}", rx_esp, rx_message, rx_crlf);
        let rx_error = format!(r"(-ERR){}({})({}){{0,1}}", rx_esp, rx_message, rx_crlf);
        let rx_disconnect = format!(r"(DISCONNECT)({}){{0,1}}", rx_crlf);
        let rx_echo= format!(r"(ECHO){}({}){}({})({}){{0,1}}", rx_esp, rx_port, rx_esp, rx_domaine, rx_crlf);
        let rx_send = format!("(SEND){}({}){}({}|{}){}({}|{}){}({})", rx_esp, rx_id_domaine, rx_esp, rx_nom_domaine, rx_tag_domaine, rx_esp, rx_nom_domaine, rx_tag_domaine, rx_esp, rx_message_interne).to_string();

        let hello = format!(r"HELLO <domaine> <random22>{}", rx_crlf).to_string();
        let connect = format!(r"CONNECT <nom-utilisateur>{}", rx_crlf).to_string();
        let param = format!(r"PARAM <round> <salt>{}", rx_crlf).to_string();
        let confirm = format!(r"CONFIRM <sha3>{}", rx_crlf).to_string();
        let register = format!(r"REGISTER <nom-utilisateur> <salt-size> <bcrypt-hash>{}", rx_crlf).to_string();
        let follow = format!(r"FOLLOW <nom-ou-tag-domaine>{}", rx_crlf).to_string();
        let msg = format!(r"MSG <message>{}", rx_crlf).to_string();
        let msgs = format!(r"MSGS <nom-domaine> <message>{}", rx_crlf).to_string();
        let ok = format!(r"+OK <message>{}", rx_crlf).to_string();
        let error = format!(r"-ERR <message>{}", rx_crlf).to_string();
        let disconnect = format!(r"DISCONNECT{}", rx_crlf).to_string();
        let echo = format!(r"ECHO <port> <domaine>{}", rx_crlf).to_string();
        let send = format!(r"SEND <id-domaine> <nom-domaine> <nom-ou-tag-domaine> <message-interne>").to_string();

        Self {
            rx_chiffre,
            rx_lettre,
            rx_lettre_chiffre,
            rx_caractere_imprimable,
            rx_crlf,
            rx_symbole,
            rx_esp,
            rx_domaine,
            rx_port,
            rx_round,
            rx_bcrypt_hash,
            rx_sha3_hex,
            rx_salt,
            rx_salt_size,
            rx_random22,
            rx_message,
            rx_message_interne,
            rx_nom_utilisateur,
            rx_tag,
            rx_nom_domaine,
            rx_tag_domaine,
            rx_id_domaine,
            rx_hello,
            rx_connect,
            rx_param,
            rx_confirm,
            rx_register,
            rx_follow,
            rx_msg,
            rx_msgs,
            rx_ok,
            rx_error,
            rx_disconnect,
            rx_echo,
            rx_send,
            hello,
            connect,
            param,
            confirm,
            register,
            follow,
            msg,
            msgs,
            ok,
            error,
            disconnect,
            echo,
            send,
        }
    }

    pub fn verify_message(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        match message.split(" ").next().unwrap().trim() {
            "HELLO" => result = self.verify_hello(message),
            "CONNECT" => result = self.verify_connect(message),
            "PARAM" => result = self.verify_param(message),
            "CONFIRM" => result = self.verify_confirm(message),
            "REGISTER" => result = self.verify_register(message),
            "FOLLOW" => result = self.verify_follow(message),
            "MSG" => result = self.verify_msg(message),
            "MSGS" => result = self.verify_msgs(message),
            "OK" => result = self.verify_ok(message),
            "ERROR" => result = self.verify_error(message),
            "DISCONNECT" => result = self.verify_disconnect(message),
            "ECHO" => result = self.verify_echo(message),
            "SEND" => result = self.verify_send(message),
            _ => {
                result.push("-ERR".to_string());
                result.push("This message is not correctly formatted".to_string());
            }
        }
        result
    }
    fn verify_hello(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_hello).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
            result.push(captures[3].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This HELLO message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_connect(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_connect).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This CONNECT message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_param(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_param).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
            result.push(captures[3].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This PARAM message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_confirm(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_confirm).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This CONFIRM message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_register(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_register).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
            result.push(captures[3].to_string());
            result.push(captures[4].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This REGISTER message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_follow(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_follow).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This FOLLOW message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_msg(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_msg).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This MSG message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_msgs(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_msgs).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
            result.push(captures[3].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This MSGS message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_ok(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_ok).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This OK message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_error(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_error).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This ERROR message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_disconnect(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_disconnect).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This DISCONNECT message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_echo(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_echo).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
            result.push(captures[3].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This ECHO message is not correctly formatted".to_string());
        }
        result
    }

    fn verify_send(&self, message: &str) -> Vec<String> {
        let mut result = Vec::new();
        let regex = Regex::new(&self.rx_send).unwrap();
        if let Some(captures) = regex.captures(message) {
            result.push(captures[1].to_string());
            result.push(captures[2].to_string());
            result.push(captures[3].to_string());
            result.push(captures[4].to_string());
            result.push(captures[5].to_string());
        } else {
            result.push("-ERR".to_string());
            result.push("This SEND message is not correctly formatted".to_string());
        }
        result
    }
}
